src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs

```diff
@@ -131,8 +131,15 @@ public void Start<TContext>(IHttpApplication<TContext> application)
 
                     return;
                 }
-                else if (!hasListenOptions)
+                else if (!hasListenOptions || (_serverAddresses.PreferHostingUrls && hasServerAddresses))
```

Comment:
I like the alternative approach; it's easier to follow the logic in that version.